src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/RefInitializationHoister.cs

```diff
@@ -0,0 +1,240 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp;
+
+internal class RefInitializationHoister<THoistedSymbolType, THoistedAccess>(SyntheticBoundNodeFactory f, MethodSymbol originalMethod, TypeMap typeMap)
+    where THoistedSymbolType : Symbol
+    where THoistedAccess : BoundExpression
+{
+    private readonly SyntheticBoundNodeFactory _factory = f;
+    private readonly MethodSymbol _originalMethod = originalMethod;
+    private readonly TypeMap _typeMap = typeMap;
+
+    internal BoundExpression? HoistRefInitialization<TArg>(
+        LocalSymbol local,
+        BoundExpression visitedRight,
+        Dictionary<Symbol, CapturedSymbolReplacement> proxies,
+        Func<TypeSymbol, TArg, LocalSymbol, THoistedSymbolType> createHoistedSymbol,
+        Func<THoistedSymbolType, TArg, THoistedAccess> createHoistedAccess,
+        TArg arg,
+        bool isRuntimeAsync)
+    {
+        Debug.Assert(
+            local switch
+            {
+                TypeSubstitutedLocalSymbol tsl => tsl.UnderlyingLocalSymbol,
+                _ => local
+            } is SynthesizedLocal
+        );
+        Debug.Assert(local.SynthesizedKind == SynthesizedLocalKind.Spill ||
+                     (local.SynthesizedKind == SynthesizedLocalKind.ForEachArray && local.Type.HasInlineArrayAttribute(out _) && local.Type.TryGetInlineArrayElementField() is object));
```

Comment:
Do we have a test for this foreach scenario on inline array?

Reply:
Will add verification in InlineArrayTests.cs